### PR TITLE
TYP: selection and groups type-hinting in groupby

### PR DESCRIPTION
**Pull request by rhshadrach**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -487,10 +496,10 @@ def __init__(
         obj: FrameOrSeries,
         keys: Optional[_KeysArgType] = None,
         axis: int = 0,
-        level=None,
+        level: Optional[IndexLabel] = None,
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py:499]: I *think* IndexLabel is too permissive here no? Would this ever just be a scalar?

File: pandas/core/groupby/grouper.py [hunk 2]
```diff
@@ -600,7 +600,7 @@ def _make_codes(self) -> None:
             self._group_index = uniques
 
     @cache_readonly
-    def groups(self) -> Dict[Hashable, np.ndarray]:
+    def groups(self) -> PrettyDict[Hashable, np.ndarray]:
```
Comment (WillAyd) [inline pandas/core/groupby/grouper.py:603]: I *think* this should just be `Dict` - while literally correct we type our API with as permissive of a type as possible for arguments, so I think should do the same on return values as well

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (WillAyd) [inline pandas/core/groupby/groupby.py:499]: `IndexLabel` is likely too permissive here - would this ever be a scalar?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/grouper.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/groupby/grouper.py:603]: Agreed on arguments, especially for user-facing methods. But does that also apply to a return, where the result is known to be a PrettyDict? In general I try to be as specific with types as possible; the more specific one can be the more issues mypy can check. I also find the specificity helps with readability.

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:499]: From the groupby docs: level : int, level name, or sequence of such, default None

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:499]: E.g. 

```
df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]}).set_index('a')
print(df.groupby(level='a').sum())
```

gives 

```
   b
a   
1  4
2  5
3  6
```

File: pandas/core/groupby/grouper.py (hunk 2 above)
Comment (WillAyd) [reply pandas/core/groupby/grouper.py:603]: The problem arises when users start using the return type in their own type checking, particularly when things are invariant. For all practical purposes (outside of the repr) this returns a Dict, so I think better downstream to label as such

Comment (rhshadrach) [reply pandas/core/groupby/grouper.py]: Thanks for the explanation, makes sense. To add to your reasons, I think PrettyDict is also not a public part of the public API. With this, it also shouldn't be leaked out via type-hints.

**Review by WillAyd — APPROVED** [review]
Lgtm @simonjayhawkins

**Comment by jreback** [comment]
can you rebase and ping on green.

**Comment by rhshadrach** [comment]
@jreback green

**Pull request merged**